### PR TITLE
Evm move

### PR DESCRIPTION
**Pull request by shemnon**
Move the EVM classes to a standalone module. This is mostly moves but
some API re-design to peel out some features not essential to the EVM,
such as privacy support and ties to the data storage subsystem.

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionProcessor.java [hunk 1]
```diff
@@ -83,24 +82,22 @@ public PrivateTransactionProcessor(
...
-      final BlockHashLookup blockHashLookup,
+      final Function<Long, Hash> blockHashLookup,
       final Bytes privacyGroupId) {
     try {
       LOG.trace("Starting private execution of {}", transaction);
 
       final Address senderAddress = transaction.getSender();
-      final EvmAccount maybePrivateSender = privateWorldState.getAccount(senderAddress);
-      final MutableAccount sender =
+      final var maybePrivateSender = privateWorldState.getAccount(senderAddress);
```
Comment (jflo) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionProcessor.java:99]: Is type inference necessary here? I'm not seeing what it is buying us here, personally I think having the type declarations here makes the following ternary operation more readable.

**Review by jflo — COMMENTED** [review]
This is great, lots of improvements. Really appreciate the clarification of instances of `worldState` vs. `worldUpdater`. The one thing I'd like to understand better is the usage of plugin-api interfaces (Hash and BlockHeader) from core.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java [hunk 2]
```diff
@@ -19,8 +19,10 @@
 
 import java.util.Optional;
 
+import org.apache.tuweni.bytes.Bytes;
+
 /** A block header capable of being processed. */
-public class ProcessableBlockHeader {
+public class ProcessableBlockHeader implements org.hyperledger.besu.plugin.data.BlockHeader {
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/ProcessableBlockHeader.java:25]: Core has always depended on plugins.  What is being changed is where the BlockHeader interface is being esposed.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionProcessor.java [hunk 3]
```diff
@@ -323,12 +330,20 @@ public TransactionProcessingResult processTransaction(
 
       final WorldUpdater worldUpdater = worldState.updater();
       final Deque<MessageFrame> messageFrameStack = new ArrayDeque<>();
+      final var contextVariablesBuilder =
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionProcessor.java:333]: It's a question of where you do the casting and how explicit the casting is.

File: evm/src/main/java/org/hyperledger/besu/evm/tracing/StandardJsonTracer.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import com.google.common.base.Joiner;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.units.bigints.UInt256;
+
+@SuppressWarnings("unused")
```
[the remark this answers is not in the document]
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/tracing/StandardJsonTracer.java:35]: We should be using it. I'll remove the annotation when I get home.  Probably a leftover from my initial move work before I deleted the other classes.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java [hunk 5]
```diff
@@ -189,27 +194,34 @@ public boolean isWorldStateAvailable(final Hash rootHash, final Hash blockHash)
           BlockHeader targetHeader = blockchain.getBlockHeader(blockHash).get();
           BlockHeader persistedHeader = maybePersistedHeader.get();
           // roll back from persisted to even with target
+          Hash persistedBlockHAsh = fromPlugin(persistedHeader.getBlockHash());
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java:197]: Tpyo

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionProcessor.java (hunk 1 above)
Comment (shemnon) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivateTransactionProcessor.java:99]: Ease of writing.  I'll adjust it.

File: ethereum/core/src/test-support/java/org/hyperledger/besu/ethereum/core/TestCodeExecutor.java [hunk 6]
```diff
@@ -72,7 +76,7 @@ public MessageFrame executeCode(
         new MessageFrameTestFixture()
             .messageFrameStack(messageFrameStack)
             .blockchain(fixture.getBlockchain())
-            .worldState(worldState)
+            .worldUpdater(worldState)
```
[the remark this answers is not in the document]
Comment (shemnon) [reply ethereum/core/src/test-support/java/org/hyperledger/besu/ethereum/core/TestCodeExecutor.java:79]: Done

File: evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java [hunk 7]
```diff
@@ -12,12 +12,14 @@
...
+package org.hyperledger.besu.evm.operation;
 
-import org.hyperledger.besu.ethereum.core.Difficulty;
-import org.hyperledger.besu.ethereum.vm.EVM;
-import org.hyperledger.besu.ethereum.vm.GasCalculator;
-import org.hyperledger.besu.ethereum.vm.MessageFrame;
+import org.hyperledger.besu.evm.EVM;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.plugin.data.Quantity;
```
[the remark this answers is not in the document]
Comment (shemnon) [reply evm/src/main/java/org/hyperledger/besu/evm/operation/DifficultyOperation.java]: I went with #2, a `BlockValues` interface.  My rationale is that there is less bit shuffling when creating a call.  Choice of Bytes is driven by future optimizations.

**Review by jflo — APPROVED** [review]

**Pull request merged**